### PR TITLE
8256298: Shenandoah: Enable concurrent stack processing

### PR DESCRIPTION
**Pull request by zhengyu123**
Please review this patch that enables concurrent stack processing for Shenandoah GC.

After this patch, all root processing is done concurrently for concurrent GC.

Test:
- [x] hotspot_gc_shenandoah Linux x86_64 and x86_32
- [x] Nightly
- [x] tier1 with -XX:+UseShenandoahGC on Linux x86_32

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8256298](https://bugs.openjdk.java.net/browse/JDK-8256298): Shenandoah: Enable concurrent stack processing

### Reviewers
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to 77582b336f37dc8797791e6bcd27e7a8f98ebb61
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2185/head:pull/2185`
`$ git checkout pull/2185`

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp [hunk 1]
```diff
@@ -506,6 +506,11 @@ class ShenandoahHeap : public CollectedHeap {
   void sync_pinned_region_status();
   void assert_pinned_region_status() NOT_DEBUG_RETURN;
 
+// ---------- Concurrent Stack Processing support
+//
+public:
+  bool uses_stack_watermark_barrier() const { return true; }
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:512]: This overrides a CollectedHeap method. I wonder if we should start adding 'override' keywords to overridden methods. However, this requires >=C++11, and might disturb backporting. Not sure about it.

File: src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2020, Red Hat, Inc. All rights reserved.
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp:2]: This is structurally similar to zStackWatermark.cpp, and perhaps 'inspired' by it. Consider adding Oracle copyright there in addition to the Red Hat copyright. We did the same in e.g. shenandoahReferenceProcessor.hpp/cpp. Also, bump the copyright year to 2021 maybe?

File: src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2020, Red Hat, Inc. All rights reserved.
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.hpp:2]: Same comment as in shenandoahStackWatermark.cpp above.

**Review by rkennke — CHANGES_REQUESTED** [review]
Wow that looks very nice and clean! I only have a few minor comments.

File: src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2021, Red Hat, Inc. All rights reserved.
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp:3]: Small nit: the original copyright in the corresponding z* files is dated 2020. Please preserve that. Leave the Red Hat copyright at 2021. Sorry.

File: src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.hpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+#include "utilities/globalDefinitions.hpp"
+
+class frame;
+class JavaThread;
+
+class ShenandoahOnStackCodeBlobClosure : public CodeBlobClosure {
+private:
+  BarrierSetNMethod* _bs_nm;
+
+  virtual void do_code_blob(CodeBlob* cb);
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.hpp:45]: There is no need to mark this virtual, or is there? I see you put override in the other place in that change, so maybe put it here too?

File: src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.hpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  ShenandoahOnStackCodeBlobClosure     _cb_cl;
+public:
+  ShenandoahStackWatermark(JavaThread* jt);
+  ThreadLocalAllocStats& stats();
+
+  static void change_epoch_id();
+private:
+  OopClosure* closure_from_context(void* context);
+  virtual uint32_t epoch_id() const;
+  virtual void start_processing_impl(void* context);
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.hpp:68]: Also, similar to the other case, we might avoid virtual here, and use override instead?

**Review by rkennke — CHANGES_REQUESTED** [review]
Better, thanks! Found a few more very minor complaints (sorry... ;-) )

**Review by rkennke — APPROVED** [review]
Looks good to me! Thank you!

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp [hunk 7]
```diff
@@ -506,6 +506,11 @@ class ShenandoahHeap : public CollectedHeap {
   void sync_pinned_region_status();
   void assert_pinned_region_status() NOT_DEBUG_RETURN;
 
+// ---------- Concurrent Stack Processing support
+//
+public:
+  bool uses_stack_watermark_barrier() const override { return true; }
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:512]: I wonder that this single use of `override` caused CLang to complain about the rest of the file: https://github.com/zhengyu123/jdk/runs/1762873416

File: src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  return _stats;
+}
+
+uint32_t ShenandoahStackWatermark::epoch_id() const {
+  return _epoch_id;
+}
+
+void ShenandoahStackWatermark::change_epoch_id() {
+  shenandoah_assert_safepoint();
+  _epoch_id ++;
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp:59]: Excess space before `++`.

**Review by shipilev — COMMENTED** [review]
Cursory review.

File: src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp (hunk 8 above)
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp]: Done

File: src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.hpp [hunk 9]
```diff
@@ -56,7 +56,7 @@ class outputStream;
   f(scan_roots,                                     "  Scan Roots")                    \
   SHENANDOAH_PAR_PHASE_DO(scan_,                    "    S: ", f)                      \
                                                                                        \
-  f(conc_mark_roots,                                "  Roots ")                        \
+  f(conc_mark_roots,                                "Concurrent Mark Roots ")          \
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.hpp:59]: This seems like fixing the regression from JDK-8255765, is it not? Should be a separate issue?

**Review by shipilev — CHANGES_REQUESTED** [review]
Good job, I have only a few minor comments.

**Review by shipilev — APPROVED** [review]

**Pull request closed without merging**